instrumentation/opentelemetry-instrumentation-kafka-python/src/opentelemetry/instrumentation/kafka/utils.py

```diff
@@ -0,0 +1,215 @@
+    def set(self, carrier: textmap.CarrierT, key: str, value: str) -> None:
+        if carrier is None or key is None:
+            return
+
+        if value:
+            value = value.encode()
+        carrier.append((key, value))
+
+
+_kafka_getter = KafkaContextGetter()
+_kafka_setter = KafkaContextSetter()
+
+
+def _enrich_span(
+    span, bootstrap_servers: List[str], topic: str, partition: int
+):
+    if span.is_recording():
+        span.set_attribute(SpanAttributes.MESSAGING_SYSTEM, "kafka")
+        span.set_attribute(SpanAttributes.MESSAGING_DESTINATION, topic)
+        span.set_attribute(SpanAttributes.MESSAGING_KAFKA_PARTITION, partition)
+        span.set_attribute(
+            SpanAttributes.MESSAGING_URL, json.dumps(bootstrap_servers)
+        )
+
+
+def _get_span_name(operation: str, topic: str):
+    return f"kafka.{operation}: {topic}"
+
+
+def _wrap_send(tracer: Tracer, produce_hook: HookT) -> Callable:
+    def _traced_send(func, instance, args, kwargs):
+        headers = KafkaPropertiesExtractor.extract_send_headers(args, kwargs)
+        if headers is None:
+            headers = []
+            kwargs["headers"] = headers
+
+        topic = KafkaPropertiesExtractor.extract_send_topic(args)
+        bootstrap_servers = KafkaPropertiesExtractor.extract_bootstrap_servers(
+            instance
+        )
+        partition = KafkaPropertiesExtractor.extract_send_partition(
+            instance, args, kwargs
+        )
+        span_name = _get_span_name("send", topic)
+        with tracer.start_as_current_span(
+            span_name, kind=trace.SpanKind.PRODUCER
+        ) as span:
+            _enrich_span(span, bootstrap_servers, topic, partition)
+            inject(
+                headers,
+                context=set_span_in_context(span),
+                setter=_kafka_setter,
+            )
+            try:
+                produce_hook(span, args, kwargs)
+            except Exception as hook_exception:  # pylint: disable=W0703
+                _LOG.exception(hook_exception)
+
+        return func(*args, **kwargs)
+
+    return _traced_send
+
+
+def _start_consume_span_with_extracted_context(
+    tracer: Tracer, headers: List, topic: str
+) -> Span:
+    extracted_context = extract(headers, getter=_kafka_getter)
+    span_name = _get_span_name("consume", topic)
+    span = tracer.start_span(
+        span_name, context=extracted_context, kind=trace.SpanKind.CONSUMER
+    )
+    new_context = set_span_in_context(span, extracted_context)
+    attach(new_context)
+    return span
+
+
+def _wrap_next(tracer: Tracer, consume_hook: HookT) -> Callable:
+    def _traced_next(func, instance, args, kwargs):
+        # End the current span if exists before processing the next record
+        current_span = trace.get_current_span()
+        if current_span.is_recording() and current_span.name.startswith(
+            "kafka.consume"
+        ):
+            current_span.end()
+
+        record = func(*args, **kwargs)
```

Comment:
Shouldn't the `func` also be inside the `with trace.use_span(span):` context? This will allow the `trace_id` to pass on to child spans

Reply:
the `func` here is the original `__next__`, that fetches the next record from the topic, it isn't a callback that process the record

Reply:
Oh, I see the usage example you provided now.
This is quite strange, is this the recommended practice with the KafkaConsumer object? It doesn't behave like a classic consumer because it doesn't run forever.
I see in the [documentation](https://kafka-python.readthedocs.io/en/master/apidoc/KafkaConsumer.html#:~:text=Subscribe%20to%20a%20list%20of%20topics%2C%20or%20a%20topic%20regex%20pattern.) there is a `subscribe` function in this object which receives a callback.

Reply:
the callback passed to `subscribe` isn't being executed on record consuming, but on partition assignment change events:
```
listener (ConsumerRebalanceListener) –
Optionally include listener callback, which will be called before and after each rebalance operation.
```
I agree it behaves strange, in the official [documentation](https://kafka-python.readthedocs.io/en/master/#kafkaconsumer)  it looks like thee consumer is meant to be used as an iterator, and I haven't find yet a way to invoke callbacks when consuming a record

Reply:
This is an interesting issue actually. Perhaps you should store the span in the instrumentor and close it on each call to `__next__`

Reply:
this solution will solve the issues of calling `__next__` from multiple consumers, but we will still have the issue of having unfinished spans when the code that process the records raises an exception, as @owais mentioned

maybe in addition to storing the span, I can also store a TTL value for it, and have some mechanism of deleting open spans when they expire

Reply:
This can be solved by adding [atexit](https://docs.python.org/3/library/atexit.html) for the garbage collector to close the span

Reply:
@oxeye-nikolay could you please elaborate on that? I thought atexit is used to call function when the interpreter exits, not an iterator or context?

Reply:
I think TTL is not a great solution for this as we don't know when a record will be realistically processed and we'll need to run something in the background. It sounds like too many moving parts and not worth it IMO.

I think we could compromise a bit on auto-instrumentation here by wrapping the actual `__next__` call in a span and providing a helper function to record processing. So each call to `__next__` automatically generates and finishes a span and represents the record being "read". In addition to this the instrumentation can provide helper context manager for let people add "processing" spans manually. Could look something like:

```python
for record in records:
    with kafka_instrumentation.process_record(record):
         # process logic here
```

this would create three spans for one record, one producer span, one consumer span and one processing span. Processing span would be child of consumer span. If user does not update the code manually then we'd only have a producer and a consumer span.

Reply:
BTW we should also instrument to have a span for the fetch/network-op that the client executes to fetch records from kafka or is that covered by the `__next__` call on the iterator implicitly?

Reply:
> @oxeye-nikolay could you please elaborate on that? I thought atexit is used to call function when the interpreter exits, not an iterator or context?

@owais You are correct! `atexit` is called when the interpreter quits. `__del__` should be used instead for this case.
Again, this will not be called when the context of the iterator exits, but rather when the `KafkaCollector` will be collected by the garbage collector.

Also, another solution would be that the iteration function of `KafkaCollector` would return a new iterator object, and then it will be collected once the code leaves the context of the iteration

Reply:
@owais regarding the fecth/network-op (i.e. the `poll` method) - this is done implicitly by calling `__next__`

regarding your solution, sounds like it can solve the issue here, but I am not in favor  of having to add the 'processing' spans manually

I'll try to think on another way, but it might be the only reasonable solution

Reply:
OK. If `__next__` includes fetching the records then we should definitely wrap that in a span. In addition to that we can think of ways to automatically instrument the processing block. We can even merge the producer and consumer parts in this PR and add processing support in a separate PR later. FWIW, nodejs Otel instrumentation for SQS also requires users to make some code changes even though they try to be as automatic as possible.

Reply:
@ItayGibel-helios how do you feel about splitting this work into two PRs. This PR could be merged if we just wrapped the `__next__` call and created a span for it. You can work on a solution to allow auto-instrumentation of the "process" block in a separate PR.

Reply:
@owais sounds good, I'll do that